### DISCUSSION TITLE
RunsOn for running benchmarks

### DISCUSSION THREAD

**Discussion by MahdiBM** [runs-on/runs-on#132] [Q&A] [answered]
We've added some benchmarks to our project so we'd want to use some stable machines, hopefully with minimum noisy-neighbors, so the benchmark results are consistent.

What do you recommend?
We're currently using `spot: false` and a `c7i` 8 CPU machine.
The results are already fine, but I thought i'd ask.

I know for benchmarks you'd want to run them on dedicated machines optimally, although it can cost.
I'm not sure if RunsOn can help with that.
I don't know if it'll be worth it to have dedicated machines on AWS. Somewhere like Hetzner can end up much cheaper I think?

**Comment by crohr** [comment, marked as answer]
Hi @MahdiBM, what kind of benchmark are you running? If you want complete isolation, you can use bare-metal instances on AWS (the ones with `-metal` suffix). They're not cheap and boot a bit slower, but then you get the full machine.

**Reply by MahdiBM** [reply]
Alright thank you.

The benchmarks for now are cpu-total-time benchmarks, I know we could run allocation benchmarks which will not require predictable cpu performance.

I'm all ears if you have any other tips though 😀.

**Reply by crohr** [reply]
I guess the most important thing is to run the bench multiple times to ensure you get stable results. Not sure it's that helpful to benchmark metal instances if you don't plan to run your real workloads on them.

**Reply by MahdiBM** [reply]
How can i choose a `c7g` metal machine?
```yml
  large-stable:
    family: "c7g.metal"
    hdd: 40
    image: ubuntu22-full-x64
    spot: false
    ssh: false
```
This doesn't work.
I can use `    family: "c7g"` probably and filter ram and CPU, but there are 2 machines with that exact name and all, so it could pick the other machine.

<img width="876" alt="Screenshot 2024-09-17 at 1 00 26 PM" src="https://github.com/user-attachments/assets/2694535b-b21e-4625-a9c8-f54ef9c8c86f">

**Reply by crohr** [reply]
It should work. What is the error you get?

**Reply by crohr** [reply]
@MahdiBM you are using an x64 image, but it is an arm instance. So use `image: ubuntu22-full-arm64` instead (assuming your code is compatible with arm).

**Reply by MahdiBM** [reply]
It should be fine on arm. Thanks again, I'll try.

**Reply by MahdiBM** [reply]
Ok thank you, this is working.
Side note for maybe future visitors: the startup time is definitely not bad, i don't think it took a minute.
I'll see with the performance and predictability.

**Reply by crohr** [reply]
Yes on arm64 it's quite fast, around 40-45s instead of 30s.

**Reply by MahdiBM** [reply]
Ok good news, the performance is incredibly predictable.
Ran a benchmark 7 times, the result always was 7.72s or 7.71s.
(c7g.metal, 3 warmup rounds, 5 iterations)
This is great.
I'll only have to check the pricing now.

**Reply by MahdiBM** [reply]
> Not sure it's that helpful to benchmark metal instances if you don't plan to run your real workloads on them.

For the record, It doesn't seem to matter considering the machine will only be used for the benchmark.
I've set the benchmark to 1 warmup round and 1 iteration. The benchmark results are basically as stable as they were with 3 warmup and 5 iteration rounds.
I think I could even remove the warmup round.
